### DISCUSSION TITLE
Backup using shell script

### DISCUSSION THREAD

**Discussion by joindevopscloud** [daws-76s/.github#6] [Q&A]
**Develop a backup script that takes a source directory and a destination directory as input and creates a compressed archive (.tar.gz) of the source directory in the destination folder with a timestamp.**

**Comment by knbob** [comment]
```
#!/bin/bash

echo "Please enter the source directory to create back: "                 #print message
read SD                                                                                              #Read input from user                                      
echo "Please enter the destination directory to save .tar.gz file: "    #print message
read DD                                                                                             #Read input from user

TIME=$(date +%F_%H_%M_%S)                                                       #Time stamp


folder=$(echo "$SD"| awk -F / '{print $NF}')                                    # to get folder name from absolute path
gzfile="$folder-$TIME.tar.gz"                                            
cd $SD
cd ..
tar -r -f "$folder-$TIME.tar.gz" $folder                                            # create .tar.gz file
mv $gzfile $DD
```
![Assignment output](https://github.com/daws-76s/.github/assets/153567059/40cee93d-2e36-421a-ad61-2d0d1b3998cf)

**Reply by Sravanthich1305** [reply]
cat 01-backup.sh
#!/bin/bash
##backup using shell script

#files list to be backup.
backup_files="/var/log/messages"

# enter the destination
dest="/home/centos/"
time=$(date +%m-%d-%y)
# Create archive filename.
archive_file=backup_messages_$time.tz

# Print start status message.
echo "Backing up $backup_files to $dest/$archive_file"
# Backup the files using tar.
tar cvzf $dest/$archive_file $backup_files
echo "Backup has completed"

#lisitng the backup file
ls -l $dest
============
output:
Backing up /var/log/messages to /home/centos//backup_messages_12-17-23.tz
tar: Removing leading `/' from member names
/var/log/messages
Backup has completed
total 4
-rw-r--r-- 1 root root 1217 Dec 17 15:24 backup_messages_12-17-23.tz

**Comment by saipdevops555** [comment]
#!/bin/bash
# Assign input arguments to variables
source_directory="$1"
destination_directory="$2"

# Check if source directory exists
if [ ! -d "$source_directory" ]; then
    echo "Error: Source directory does not exist."
    exit 1
fi

# Check if destination directory exists, create if not
if [ ! -d "$destination_directory" ]; then
    mkdir -p "$destination_directory"
fi

# Create a timestamp for the backup file
timestamp=$(date +"%Y%m%d%H%M%S")

  # Compress and archive the source directory
  backup_filename="${destination_directory}/backup_${timestamp}.tar.gz"
tar -czvf "$backup_filename" "$source_directory"

# Check if the backup was successful
if [ $? -eq 0 ]; then
    echo "Backup successful. Archive saved as: $backup_filename"
else
    echo "Error: Backup failed."
fi


**Explaination: Here's what each part of this line does -  backup_filename="${destination_directory}/backup_${timestamp}.tar.gz":**

${destination_directory}: This is a variable representing the destination directory for the backup.

/: This is the directory separator in Unix-like systems.

backup_: This is a prefix for the backup filename.

${timestamp}: This is another variable representing the timestamp. It seems like you're planning to include the timestamp in the filename, which is a common practice to ensure uniqueness.

.tar.gz: This is a file extension commonly used for compressed tar archives.

Putting it all together, this line creates a backup filename composed of the destination directory, the prefix "backup_", the timestamp, and the file extension ".tar.gz".

For example, if destination_directory is /path/to/backups and timestamp is 20231215120000 (assuming a timestamp format of YYYYMMDDHHMMSS), the resulting backup_filename would be /path/to/backups/backup_20231215120000.tar.gz.

**Explaination for tar -czvf "$backup_filename" "$source_directory"**

compress and archive a source directory in a Unix-like shell (such as Bash), you can use the tar
Explanation of the options used:

-c: Create a new archive.
-z: Compress the archive using gzip.
-v: Verbosely list the files processed.
-f: Specify the archive file name.

**Comment by Bhanuprakash-1995** [comment]
#!/bin/bash
source_folder=$1
destination_folder=$2
current_date=$(date +"%Y-%m-%d") # Store the current date in the 'current_date' variable

if [ ! -d "$source_folder" ]; then
    echo -e "$source_folder does not exist"
    exit 1
fi

if [ ! -d "$destination_folder" ]; then
    echo -e "$destination_folder does not exist"
else
    mkdir -p "$destination_folder"
fi

tar -czvf "$destination_folder/${current_date}.tar.gz" "$source_folder"
